gobblin-hive-registration/src/main/java/gobblin/hive/HiveRegister.java

```diff
@@ -0,0 +1,302 @@
+      log.warn("Hive database " + dbName + " already exists");
+    } catch (TException e) {
+      log.error("Unable to create Hive database " + dbName);
+      throw Throwables.propagate(e);
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  private Table getOrCreateTable(String dbName, String tableName, Optional<HivePartition> partition,
+      StorageDescriptor sd) {
+    Lock lock = locks.get(dbName + tableName);
+    lock.lock();
+
+    try {
+      return this.client.getTable(dbName, tableName);
+    } catch (NoSuchObjectException e) {
+      return createTable(dbName, tableName, partition, sd);
+    } catch (TException e) {
+      log.error(String.format("Unable to check existence of Hive table %s in db %s", tableName, dbName));
+      throw Throwables.propagate(e);
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  private Table createTable(String dbName, String tableName, Optional<HivePartition> partition, StorageDescriptor sd) {
+    Table table = new Table();
+    table.setParameters(Maps.<String, String> newHashMap());
+    table.getParameters().put("EXTERNAL", Boolean.TRUE.toString());
+    table.setDbName(dbName);
+    table.setTableName(tableName);
+    if (partition.isPresent()) {
+      table.setPartitionKeys(partition.get().getKeys());
+    }
+
+    Preconditions.checkState(this.hiveOwner.isPresent(), "Missing required property " + HiveRegProps.HIVE_OWNER);
+    table.setOwner(this.hiveOwner.get());
+    table.setTableType("EXTERNAL_TABLE");
+    table.setSd(sd);
+    try {
+      this.client.createTable(table);
+    } catch (TException e) {
+      log.error(String.format("Unable to create table %s from db %s", tableName, dbName));
+      throw Throwables.propagate(e);
+    }
+    return table;
+  }
+
+  public void deregister(String dbName, String tableName) {
+    Lock lock = locks.get(dbName + tableName);
+    lock.lock();
+
+    try {
+      if (this.client.tableExists(dbName, tableName)) {
+        this.client.dropTable(dbName, tableName);
+      }
+    } catch (TException e) {
+      log.error(String.format("Unable to deregister table %s from db %s", tableName, dbName));
+      throw Throwables.propagate(e);
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  public void deregister(String dbName, String tableName, List<String> partitionValues) {
+    Lock lock = locks.get(dbName + tableName);
+    lock.lock();
+
+    try {
+      client.dropPartition(dbName, tableName, partitionValues, false);
+    } catch (NoSuchObjectException e) {
+      // Partition does not exist. Nothing to do
+    } catch (TException e) {
+      log.error(String.format("Unable to check existence of Hive partition %s in table %s in db %s", partitionValues,
+          tableName, dbName));
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  private void addPartition(Table table, HivePartition partition) throws TException {
+    Preconditions.checkArgument(table.getPartitionKeysSize() == partition.getValues().size());
```

Comment:
This method takes a `Table` and partition names and values. The actual `Partition` object is constructed in this method using the dbname, tablename and serde of the given `Table`.